README.md
# WThermostat - FAS-Fork

[![GitHub version](https://img.shields.io/github/release/fashberg/WThermostatBeca.svg)](https://github.com/fashberg/WThermostatBeca/releases/latest)
[![GitHub release (latest by date including pre-releases)](https://img.shields.io/github/v/release/fashberg/WThermostatBeca?include_prereleases&label=beta)](https://github.com/fashberg/WThermostatBeca/releases)
[![GitHub download](https://img.shields.io/github/downloads/fashberg/WThermostatBeca/total.svg)](https://github.com/fashberg/WThermostatBeca/releases/latest)
[![Gitpod ready-to-code](https://img.shields.io/badge/Gitpod-ready--to--code-blue?logo=gitpod)](https://gitpod.io/#https://github.com/fashberg/WThermostatBeca)
[![Github Actions](https://img.shields.io/github/workflow/status/fashberg/WThermostatBeca/Build)](https://github.com/fashberg/WThermostatBeca/actions)

Fork of <https://github.com/klausahrenberg/WThermostatBeca> with some new features.

Replaces original Tuya firmware on Beca/Moes/Qiumi thermostat with ESP8266 WiFi module.

## Supported Devices

The firmware has been tested or reported to work with following devices:

* BHT-002 WiFi Thermostat
  * BHT-002-GALW (Water/Gas heating)
  * BHT-002-GBLW (Electrical floor heating)
  * BHT-002-GCLW (Water/Gas Boiler)
* BHT-6000 WiFi Thermostat
  * BHT-6000-GALW, BHT-6000-GBLW, BHT-6000-GCLW
* BHT-3000 WiFi Thermostat
  * BHT-3000-GALW, BHT-3000-GBLW, BHT-3000-GCLW
  * also known as BTH-3000L-ABLW, BTH-3000L-GBLW, BTH-3000L-GCLW
* BAC-1000-ALW WiFi Thermostat for Conditioning (heating, cooling, ventilation)
  * also known as BAC-002-ALW

![homeassistant](docs/images/bac-002-wifi.jpg)

## Features

* No Cloud dependencies!
* Enables thermostat to communicate via MQTT and/or Mozilla WebThings
* Autodiscovery for WebThings via mDNS
* Configuration of connection, device parameters and schedules via web interface
* Provides NTP, time zone handling and Daylight-Saving-Calculation to set the clock of thermostat
* Provides Fallback to Access Point mode if requested using panel-buttons
* Reading and setting of all parameters via MQTT
* Reading and setting of main parameters via WebThings
* Only BHT-002-GxLW: actualFloorTemperature (external temperature sensor)
* Only BAC-002-ALW: fanSpeed:auto|low|medium|high; systemMode:cooling|heating|ventilation
* Reading and setting of time schedules via MQTT
* Reading and setting of Schedules via Web-GUI
* Automatic switching back to temperature-scheduling from manual mode next schedule interval (optional)
* **FAS-Fork enhancements**
  * RelayStateCalculation - shows current state of heating/cooling - without soldering _(fas)_
  * Autodiscovery for Home Assistant via MQTT _(fas)_
  * Logging to MQTT _(fas)_
  * Logging to WebBrowser using Websockets _(fas)_
  * Captive Portal support for setup _(fas)_
  * show password in cleartext for setup _(fas)_
  * TimeBug Workaround (precicion fixed to 1 sec) _(fas)_
  * All State Information at status page _(fas)_
  * GitPod-Support _(fas)_
  * PlatformIO Direct OTA-Upload to device _(fas)_
  * Eco-Mode fixed to 20 degrees _(fas)_
  * RSSI Information (Info-Page, MQTT and Home-Assistant Support) _(fas)_
  * More detailed Documentation _(fas)_

_(fas)_: Only available in -fas version

## Hardware-Versions

You need the WiFi Version! (W in product name suffix, like -GAL**W**). There are also a versions without WLAN (Offline-Version and ZigBee-Version (-GALZBW)). See [Diagnose Can we add WiFi to Non-WiFi Thermostat bac-002.pdf](./docs/Diagnose%20Can%20we%20add%20WiFi%20to%20Non-WiFi%20Thermostat%20bac-002.pdf)

The BHT Version is for heating only. The BAC-Version has modes Cooling, Heating and Ventilation.
The BHT-002-GA/GB/GC versions only differs in relays-wiring.

### Difference between GA, GB and GC

* GA - Water-Heating
  * Two Relays for opening and closing valve
  * Only one Relay will be closed at the same time
  * Closing Relay PIN 1 - PIN 3 (N or L)
  * Opening Relay PIN 2 - PIN 3 (N or L)
  * Product Spec says Max Power: 3 A
* GB - Electric-Heating
  * Connect Heating between PIN 1 and PIN 2
  * Product Spec says Max Power: 16 A
* GC - Water/Gas Boiler
  * One Relay - potential free (dry contact)
  * Relay on PIN 1 - PIN 2 (dry contacts)
  * Product Spec says Max Power: 3 A

### Hardware Installation

Here a description of how to install the hardware - indepentdent from Open Source WThermostat.
**No special Hardware installation necessary for Upgrading to this firmware!**

* see [Installation.md](Installation.md)

## Download binaries

Pre-built binaries can be downloaded at [releases-page](https://github.com/fashberg/WThermostatBeca/releases).

## Software Installation

:warning: **Nearly all devices are now shipped with a new Firmware which cannot be upgraded Over The Air with non-tuya-servers.**
 In this case you cannot longer use tuya-convert and you have to open the device for soldering or use a 3d-printed pogo-adapter.

You can install the firmware either

* by using tuya-convert - no Hardware Modifications necessary
* by soldering cables to the ESP-Modules using an ESP/Arduino-Programmer (3,3 Volt TTL)
* by using OTA Upgrade on tasmota or other Open Source ESP Firmware

Yoube video of converting to 1.14-fas with tuya-convert: <https://youtu.be/fqfghJqnK_8>

Follow instructions here: [Flashing.md](Flashing.md)

## Initial configuration

* The devices becomes an Access Point
* Connect to WiFi `Thermostat-xxxxxx` with Password `12345678`
* Open `http://192.168.4.1` or wait for Captive Portal to be opened on Android
* Configure your WiFi Settings
* After initial setup, the device configuration is available at `http://<device_ip>/`
* Configure device model, clock and other settings

See [Configuration.md](Configuration.md) for a detailed explanation of the setup.

## Upgrading

You can Upgrade over the Air from other versions of WThermostatBeca on the Web-GUI.
Please check all settings after Upgrade!

If you upgrade from Klaus-Ahrenberg's version, please note:

* Only WLAN settings are kept
* If you get "Not Enough Space" you can use WThermostat_1.xx-fas-minial.bin as interim version.
  * WLAN settings are kept
  * no MQTT, no Beca-control, no Clock in -minimal
  * Then Upgrade to normal version
  * Not necessary when upgrading from fas to fas versions

## Screenshots

Main Screen:

![setup_main](docs/images/Setup_Main.png)  

Network Screen:

![setup_network](docs/images/Setup_Network.png)  

Schedule Screen:

![setup_schedules](docs/images/Setup_Schedules.png)  

Live WebLog Screen:

![weblog_screen](docs/images/weblog.png)  

## Integration in Home Assistant

![homeassistant](docs/images/homeassistant.png)
![homeassistant](docs/images/hass_detail.png)
![hass_discovery](docs/images/hass_discovery.png)

### Controlling BAC in HomeAssistant

* Use Operation to Switch between 'heat', 'cool', 'fan_only' and 'off'
* Use Preset to switch between 'None' (manual), 'Scheduler' and 'eco'
* Changing temp disables scheduler and switches to preset 'None'

![homeassistant](docs/images/hass-bac.png)

### HASS Autodiscovery

WThermostatBeca supports optional HASS-Autodiscovery since Version 1.08-fas.
You must enable it at Thermostat (settings network) and of course it must be enabled in your HASS configuration.yaml file:

```yaml
mqtt:
  broker: <ip of broker>
  discovery: true
  discovery_prefix: homeassistant
```

#### Use a persistent MQTT-Broker

You should use a persistent MQTT-Broker, keeping all retained messages during reboot/restart.
HASS integrated MQTT broker is not persistent.
Otherwise all autodiscovered messages are getting lost and you must reboot all devices, to send HASS autodiscover messages again!
See: <https://www.home-assistant.io/docs/mqtt/discovery/>

Example for users of mosquitto, file mosquitto.conf:

```sh
persistence true
persistence_file mosquitto.db
persistence_location /var/lib/mosquitto/
autosave_interval 1800
# The default if not set is to never expire persistent clients.
persistent_client_expiration 24h

```

See <https://mosquitto.org/man/mosquitto-conf-5.html> for more details.

### HASS Manual Configuration

For manual Configuration here is an example for your configuration.yaml file:

**Heating:**

```yaml
climate:
  - platform: mqtt
    name: Room_Thermostat
    availability_topic: "home/room/tele/LWT"
    payload_available: "Online"
    payload_not_available: "Offline"
    action_topic: "home/room/stat/things/thermostat/properties"
    action_template: "{{value_json.action}}"
    temperature_command_topic: "home/room/cmnd/things/thermostat/properties/targetTemperature"
    temperature_state_topic: "home/room/stat/things/thermostat/properties"
    temperature_state_template: "{{value_json.targetTemperature}}"
    current_temperature_topic: "home/room/stat/things/thermostat/properties"
    current_temperature_template: "{{value_json.temperature}}"
    mode_command_topic: "home/room/cmnd/things/thermostat/properties/mode"
    mode_state_topic: "home/room/stat/things/thermostat/properties"
    mode_state_template: "{{value_json.mode}}"
    modes: [ "heat", "auto", "off" ]
    preset_mode_command_topic: "home/room/cmnd/things/thermostat/properties/preset"
    preset_mode_state_topic: "home/room/stat/things/thermostat/properties"
    preset_mode_value_template: "{{value_json.preset}}"
    preset_modes: ["away"]
    min_temp: 5
    max_temp: 35
    temp_step: 0.5
    precision: 0.5
```

**Air Conditioning:**

The BAC-002/BAC-1000 does not support auto mode. The Device (MCU) can not switch automatically between heating, cooling or fan mode.

To Switch between Automatic Scheduling and Manual Mode we use 'holdState' with the following mode: 'scheduler' and 'manual', combined with mode 'eco'.
If you choose 'None' it disables 'eco' and device jumps back to last mode (scheduler/manual).
Do not use away_mode_ combined with holdState.

```yaml
climate:
  - platform: mqtt
    name: Bedroom AC Control
    availability_topic: "home/bedroom/tele/LWT"
    payload_available: "Online"
    payload_not_available: "Offline"
    action_topic: "home/bedroom/stat/things/thermostat/properties"
    action_template: "{{value_json.action}}"
    temperature_command_topic: "home/bedroom/cmnd/things/thermostat/properties/targetTemperature"
    temperature_state_topic: "home/bedroom/stat/things/thermostat/properties"
    temperature_state_template: "{{value_json.targetTemperature}}"

Changes.md
# Changelog

## Version 1.23-fas

* Support for static IP-Addresses - thanks to @N-b-dy
* Support for current IP-Address sensor for Home Assistant - thanks to @Henry-Sir
* Fix for new Home Assistant Autodiscovery format  (fixes [#205](https://github.com/fashberg/WThermostatBeca/issues/205))

## Version 1.22-fas

* Fixes invalid JSON responses/escaping (fixes #153)
* Fixes MQTT-Reconnect
* Doc updates

## Version 1.21-fas

* reimplemented Away-Mode with Preset (Heating and A/C)
* fixed scheduler-Mode with Preset 'Scheduler' (A/C)
* updated docs and samples

## Version 1.20-fas

### Changes in this Release

* HA 2022.9 compatibility #153, #146. Thanks to @labodj
* Update to latest dependencies and compatibility @labodj/@fashberg
* decreased minimum temp by @zawasp  / #98
* removed away-mode @jressel01 #148

### And Build CI updates

* moved from travis-ci to github-actions @fashberg

## Version 1.19-fas

* Fix/Enhancement #54: If enabled, all single Properties are sent out every notification interval, not only if changed
* Fix #52: Documentation on SSID and renamed Wifi to WiFi
* Fix #53 and #64: Boolean Values wrong

## Version 1.18-fas

* Fixed Relay State Calculation #26 again
* Daylight Saving Bug #62 fixed

## Version 1.17-fas

* Fix Relay State Calculation #26 - thanks IATkachenko
* Fix reported targetTemperature for Schdule Mode #50

## Version 1.16-fas

* NewFeature: RelayStateCalculation - shows current state of heating/cooling
* NewFeature: WLAN Signal Level (RSSI) visible (MQTT, Info-Page, HASS)
* NewFeature: Full Support for BAC-002 with Home-Assistant autodiscovery
* HASS Climate Autodiscovery BAC-002
  * removed 'auto' from Read/Write-Property ``mode`` (not supported by device). Valid Modes: ``["heat","cool","fan_only","off"]``
  * Read/Write-Property ``holdState`` with possible values ``["scheduler", "manual", "away"]`` - switches between Scheduler, Manual and Eco-Mode
  * Fixed fanMode behaviour (auto-mode now working, setting fan does not switches to fan_only)
  * Fixed manual configuration snippet in README.md
* NewFeature: Live WebLog in Browser (WebSockets)
* NewFeature: Workaround for imprecise clock on LCD
* NewFeature: All Information on Info-Page
* Develpment: Gitpd support (<https://github.com/fashberg/WThermostatBeca#cloud-development-using-gitpod>)
* Develpment: Travis CI support (<https://travis-ci.org/github/fashberg/WThermostatBeca>)
* Develpment: platformio_override.ini documentation and sample
* Small Fixes/Enhancements
  * ecoMode temperature Fixed to 20 degrees (not changeable in MCU) #32
  * WebThings autodiscovery fix
  * IceBucketChallenge: Support for negative temperatures
  * Apple Captive Portal support
  * Case Insensitive true/false (HomeAssistant Uses "True" if boolean values are enabled) #32
  * Typos

## Version 1.15-fas

* Bugfix issue #20: Hardware Hacks not saved (thanks ilkkaolavas)
* Bugfix: no instant mqtt reconnect after network-reset mode
* New: floorSensor configurable (default off). If Off: not shown in MQTT/WebThings
  * If enabled und HASS-Autodiscovery also enabled: show additional Sensor in HASS
* New: Support changing switchBackToAuto via MQTT
  * saved in EEPROM permanently on Change
  * mosquitto_pub -h mqtt -t TOPIC/cmnd/things/thermostat/properties/switchBackToAuto  -m "true"
* Bugfix: Crash on setting non-existant property using MQTT
* Bugfix: No default clock settings after fresh install
* Improvement: Show empty Passwordfield if password is empty instead of \_\_NOCHANGE__
* by Folke Ashberg <folke@ashberg.de>

## Version 1.14-fas

* MQTT Last Will & Testament (topic/tele/LWT)
* MQTT Hass Availibility Topic Support (also with autodiscover)
* Support for switching back automatically from manual temperature to scheduled temperature
* Reading NetworkSettings support for 1.00-1.11 and 1.xx-fas Versions
* Writing NetworkSettings same way Version >=1.09 does
* Changed '_' in default hostname to '-' to be RFC952 compliant
* Changed default Hostname to "Thermostat" (to fit hostname in 16 chars)
* Show passwords fields in ClearText on Request
* by Folke Ashberg <folke@ashberg.de>

## Version 1.13-fas

* Never use thirteen (tm)

## Version 1.12-fas

* HASS Autodiscovery interval fix #10 (thanks again austwhite)
* installation documentation
* by Folke Ashberg <folke@ashberg.de>

## Version 1.11-fas

* HASS Autodiscovery Fix and Improvements
* by <https://github.com/austwhite>
* platform.io automatic firmware naming
* documentation updates
* fixed targetTemp wrong on change from scheduleMode to Manual #9 (thanks austwhite)
* by Folke Ashberg <folke@ashberg.de>

## Version 1.10-fas

* introduced -minimal version for interim-upgrade on small sketch size
* new "action" property for hass - thanks austwhite
* keep only WLAN settings from compatibility Firmware
* Fix too less page-buffor for startpage (reboot button missing)
* Uptime seconds fix
* by Folke Ashberg <folke@ashberg.de>

## Version 1.09-fas

* WebThings back working (/things) - thanks erelor
* Optional sending of changed values to its own MQTT topic with no JSON - thanks Bettman66
* by Folke Ashberg <folke@ashberg.de>

## Version 1.08-fas


WThermostat/WLogDevice.h
#ifndef LOGDEVICE_H
#define	LOGDEVICE_H

#include "Arduino.h"
#include <ESP8266WiFi.h>
#include <EEPROM.h>
#include "../lib/WAdapter/WAdapter/WDevice.h"

const char* ID_LOGDEVICE PROGMEM = "logging";
const char* NAME_LOGDEVICE PROGMEM = "Logging";

const char* PROP_LOGLEVELBYTE PROGMEM = "logLevelByte";
const char* PROP_LOGLEVEL PROGMEM = "logLevel";
const char* TITL_LOGLEVEL PROGMEM = "logLevel";
const char* PROP_LOGLEVELWEB PROGMEM = "logLevelWeb";
const char* PROP_LOGLEVELWEBBYTE PROGMEM = "logLevelWebByte";

const char* PROP_LOGRES1 PROGMEM = "logres1";
const char* PROP_LOGRES2 PROGMEM = "logres2";
const char* PROP_LOGRES3 PROGMEM = "logres3";
const char* PROP_LOGRES4 PROGMEM = "logres4";

const char* ID_PAGE_LIVEWEBLOG PROGMEM = "weblog";
const char* TITLE_PAGE_LIVEWEBLOG PROGMEM = "Live WebLog";

const char* LOG_MODE_SILENT  PROGMEM = "silent";
const char* LOG_MODE_FATAL   PROGMEM = "fatal";
const char* LOG_MODE_ERROR   PROGMEM = "error";
const char* LOG_MODE_WARNING PROGMEM = "warning";
const char* LOG_MODE_NOTICE  PROGMEM = "notice";
const char* LOG_MODE_TRACE   PROGMEM = "trace";
const char* LOG_MODE_VERBOSE PROGMEM = "verbose";


const char* URI_WSLOG PROGMEM = "/wslog";

const static char HTTP_WEBLOG[]         PROGMEM = R"=====(
	<textarea readonly id="t1" cols="340" rows="25" wrap="off" name="t1" style="resize:vertical;width:98%;height:318px;padding:5px;overflow:auto;background:#1f1f1f;color:#65c115"></textarea><br>
	Hints: Change LogLevel <a href="/logging" target="_l">here</a> and reinitialize Thermostat <a href="/thermostat_reinit" target="_l">there</a>. Use LogLevel Verbose to see all MCU communication.
)=====";



const static char HTTP_WEBLOG_SCRIPT[]             PROGMEM = R"=====(<script>
	var ws = null;
	var pt = null;
	function ping() {
		if (!ws) return;
		if (ws.readyState !== 1) return;
		ws.send("heartbeat");
		pt=setTimeout(ping, 3000);
	}
	function msg(m){
		eb('t1').value+=m;
		eb('t1').scrollTop=99999;
	}
	function wsinit(){
		try {
			var url = "ws://" + location.host + "/wslog";
			msg('Connecting to ' + url + '\n');
			ws = new WebSocket(url, ["log"]);
		} catch(e){
			msg('Failed to connect\n');
			setTimeout(function(){ wsinit(); }, 3000);
			return;
		}
		ws.onopen = function(){
			msg('Connected\n');
			if (pt) clearTimeout(pt);
			pt=setInterval(function(){ping();}, 3000);
		}
		ws.onmessage = function(e){
			msg(e.data+'\n');
		}
		ws.onclose = function(e){
			msg('WebSocket Connection Closed! '+e.reason+'\n\n');
			delete ws;
			setTimeout(function(){ wsinit(); }, 3000);
		}
		ws.onerror = function(err) {
			msg('WebSocket Error! \n\n');
			//ws.close();
		}
	}
	wsinit();
</script>
)=====";

AsyncWebSocket wslog(URI_WSLOG);

class WLogDevice: public WDevice {
public:

    WLogDevice(WNetwork* network) 
        	: WDevice(network, ID_LOGDEVICE, NAME_LOGDEVICE, network->getIdx(), DEVICE_TYPE_LOG) {
		
		this->providingConfigPage = true;
		this->configNeedsReboot = false;
		this->mainDevice = false;
		this->visibility = MQTT;

		/* properties */

		if (network->getSettingsOld()){
			if (network->getSettingsOld()->getNetworkSettingsVersion()==NETWORKSETTINGS_PRE_FAS114){
				network->log()->notice(F("Reading WLogSettings PRE_FAS114"));
				network->getSettingsOld()->setByte(PROP_LOGLEVELBYTE, LOG_LEVEL_WARNING);
			} else if (network->getSettingsOld()->getApplicationSettingsVersion()==network->getSettingsOld()->getApplicationSettingsCurrent()-1){
				network->log()->notice(F("Reading WLogSettings FLAG_OPTIONS_APPLICATION -1"));
				network->getSettingsOld()->setByte(PROP_LOGLEVELBYTE, LOG_LEVEL_WARNING);
			}
		}

		logLevelByte=nullptr;
		logLevelWebByte=nullptr;
		this->logLevelByte = network->getSettings()->setByte(PROP_LOGLEVELBYTE,
			(network->getSettingsOld() && network->getSettingsOld()->existsSetting(PROP_LOGLEVELBYTE) ? network->getSettingsOld()->getByte(PROP_LOGLEVELBYTE) : LOG_LEVEL_WARNING));
		this->logLevelByte->setVisibility(NONE);
    	this->addProperty(logLevelByte);
		setlogLevelByte(constrain(getlogLevelByte(),LOG_LEVEL_SILENT, LOG_LEVEL_VERBOSE ));

		this->logLevelWebByte = network->getSettings()->setByte(PROP_LOGLEVELWEBBYTE,
			(network->getSettingsOld() && network->getSettingsOld()->existsSetting(PROP_LOGLEVELWEBBYTE) ? network->getSettingsOld()->getByte(PROP_LOGLEVELWEBBYTE) : LOG_LEVEL_TRACE));
		setlogLevelWebByte(LOG_LEVEL_TRACE);
		this->logLevel = new WProperty(PROP_LOGLEVEL, TITL_LOGLEVEL, STRING, 8);
		this->logLevel->setAtType(PROP_LOGLEVEL);
		this->logLevel->addEnumString(LOG_MODE_SILENT);
		this->logLevel->addEnumString(LOG_MODE_FATAL);
		this->logLevel->addEnumString(LOG_MODE_ERROR);
		this->logLevel->addEnumString(LOG_MODE_WARNING);
		this->logLevel->addEnumString(LOG_MODE_NOTICE);
		this->logLevel->addEnumString(LOG_MODE_TRACE);
		this->logLevel->addEnumString(LOG_MODE_VERBOSE);
		this->logLevel->setVisibility(MQTT);
		this->addProperty(logLevel);
		this->logLevelWeb = new WProperty(PROP_LOGLEVEL, TITL_LOGLEVEL, STRING, 8);
		this->logLevelWeb->setAtType(PROP_LOGLEVELWEB);
		this->logLevelWeb->addEnumString(LOG_MODE_SILENT);
		this->logLevelWeb->addEnumString(LOG_MODE_FATAL);
		this->logLevelWeb->addEnumString(LOG_MODE_ERROR);
		this->logLevelWeb->addEnumString(LOG_MODE_WARNING);
		this->logLevelWeb->addEnumString(LOG_MODE_NOTICE);
		this->logLevelWeb->addEnumString(LOG_MODE_TRACE);
		this->logLevelWeb->addEnumString(LOG_MODE_VERBOSE);
		this->logLevelWeb->setVisibility(MQTT);
		this->addProperty(logLevelWeb);

		// some reserved Bytes
		network->getSettings()->setByte(PROP_LOGRES1, 255);
		network->getSettings()->setByte(PROP_LOGRES2, 255);
		network->getSettings()->setByte(PROP_LOGRES3, 255);
		network->getSettings()->setByte(PROP_LOGRES4, 255);

		// first apply stored Byte value from EEPROM
		this->setlogLevelMqtt(getlogLevelByte());
		this->setlogLevelWeb(getlogLevelWebByte());
		// then set Handler, because now logLevelByte follows String logLevel
		this->logLevel->setOnChange(std::bind(&WLogDevice::onlogLevelMqttChange, this, std::placeholders::_1));
		this->logLevelWeb->setOnChange(std::bind(&WLogDevice::onlogLevelWebChange, this, std::placeholders::_1));

		// Pages		
		WPage * weblogPage=new WPage(ID_PAGE_LIVEWEBLOG, TITLE_PAGE_LIVEWEBLOG);
		weblogPage->setPrintPage([this,weblogPage](AsyncWebServerRequest *request, AsyncResponseStream* page) {
			this->network->log()->notice(PSTR("weblog"));
			page->print(FPSTR(HTTP_WEBLOG));
			page->print(FPSTR(HTTP_HOME_BUTTON));
			page->print(FPSTR(HTTP_WEBLOG_SCRIPT));
		});
		this->addPage(weblogPage);
    
    }

    virtual void printConfigPage(AsyncWebServerRequest *request, AsyncResponseStream* page) {
    	network->log()->notice(F("Log config page"));
    	page->printf_P(HTTP_CONFIG_PAGE_BEGIN, getId());
    	//ComboBox with model selection

		page->printf_P(HTTP_COMBOBOX_BEGIN, F("MQTT Log Mode:"), "lm");
		page->printf_P(HTTP_COMBOBOX_ITEM, "0", (getlogLevelByte() == LOG_LEVEL_SILENT  ? HTTP_SELECTED : ""), F("Logging Disabled"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "1", (getlogLevelByte() == LOG_LEVEL_FATAL   ? HTTP_SELECTED : ""), F("Fatal Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "2", (getlogLevelByte() == LOG_LEVEL_ERROR   ? HTTP_SELECTED : ""), F("Error Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "3", (getlogLevelByte() == LOG_LEVEL_WARNING ? HTTP_SELECTED : ""), F("Warning Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "4", (getlogLevelByte() == LOG_LEVEL_NOTICE  ? HTTP_SELECTED : ""), F("Notice Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "5", (getlogLevelByte() == LOG_LEVEL_TRACE   ? HTTP_SELECTED : ""), F("Trace Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "6", (getlogLevelByte() == LOG_LEVEL_VERBOSE ? HTTP_SELECTED : ""), F("Verbose Messages"));
		page->printf_P(HTTP_COMBOBOX_END);

		page->printf_P(HTTP_COMBOBOX_BEGIN, F("Web Log Mode:"), "lw");
		page->printf_P(HTTP_COMBOBOX_ITEM, "0", (getlogLevelWebByte() == LOG_LEVEL_SILENT  ? HTTP_SELECTED : ""), F("Logging Disabled"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "1", (getlogLevelWebByte() == LOG_LEVEL_FATAL   ? HTTP_SELECTED : ""), F("Fatal Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "2", (getlogLevelWebByte() == LOG_LEVEL_ERROR   ? HTTP_SELECTED : ""), F("Error Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "3", (getlogLevelWebByte() == LOG_LEVEL_WARNING ? HTTP_SELECTED : ""), F("Warning Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "4", (getlogLevelWebByte() == LOG_LEVEL_NOTICE  ? HTTP_SELECTED : ""), F("Notice Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "5", (getlogLevelWebByte() == LOG_LEVEL_TRACE   ? HTTP_SELECTED : ""), F("Trace Messages"));
		page->printf_P(HTTP_COMBOBOX_ITEM, "6", (getlogLevelWebByte() == LOG_LEVEL_VERBOSE ? HTTP_SELECTED : ""), F("Verbose Messages"));
		page->printf_P(HTTP_COMBOBOX_END);

    	page->printf_P(HTTP_CONFIG_SAVE_BUTTON);
		page->printf_P(HTTP_HOME_BUTTON);
    }
    void saveConfigPage(AsyncWebServerRequest *request) {
		network->log()->notice(F("Log Beca config save lm=%s/%d"), getValueOrEmpty(request, "lm").c_str(), getValueOrEmpty(request, "lm").toInt());
		setlogLevelByte(constrain(getValueOrEmpty(request, "lm").toInt(),LOG_LEVEL_SILENT, LOG_LEVEL_VERBOSE ));
		this->setlogLevelMqtt(getlogLevelByte());
		network->log()->notice(F("Log Beca config save lw=%s/%d"), getValueOrEmpty(request, "lw").c_str(), getValueOrEmpty(request, "lw").toInt());
		setlogLevelWebByte(constrain(getValueOrEmpty(request, "lw").toInt(),LOG_LEVEL_SILENT, LOG_LEVEL_VERBOSE ));
		this->setlogLevelWeb(getlogLevelWebByte());
    }

    void loop(unsigned long now) {
		wslog.cleanupClients();
    }

    void handleUnknownMqttCallback(String stat_topic, String partialTopic, String payload, unsigned int length) {
		//logCommand(((String)"handleUnknownMqttCallback " + stat_topic + " / " + partialTopic + " / " + payload).c_str());
    }

	byte getlogLevelByte() {
		if (!logLevelByte) return LOG_LEVEL_SILENT;
		return logLevelByte->getByte();
	}
	byte getlogLevelWebByte() {
		if (!logLevelWebByte) return LOG_LEVEL_SILENT;
		return logLevelWebByte->getByte();
	}
	void setlogLevelByte(byte lm) {
		network->log()->notice(F("WLogDevice setlogLevelByte (%d)"), lm);
		logLevelByte->setByte(lm);
		setMinLogLevel();
	}
	void setlogLevelWebByte(byte lw) {
		network->log()->notice(F("WLogDevice setlogLevelWebByte (%d)"), lw);
		logLevelWebByte->setByte(lw);
		setMinLogLevel();
	}

	void setMinLogLevel(){
		unsigned int toSet=LOG_LEVEL_SILENT;
		if (getlogLevelByte() > toSet) toSet=getlogLevelByte();
		if (getlogLevelWebByte() > toSet) toSet=getlogLevelWebByte();
		network->log()->setLevelNetwork(toSet);

	}

    void onlogLevelMqttChange(WProperty* property) {
		if (property->equalsString(LOG_MODE_SILENT)) setlogLevelByte(LOG_LEVEL_SILENT);
		else if (property->equalsString(LOG_MODE_FATAL)) setlogLevelByte(LOG_LEVEL_FATAL);
		else if (property->equalsString(LOG_MODE_ERROR)) setlogLevelByte(LOG_LEVEL_ERROR);
		else if (property->equalsString(LOG_MODE_WARNING)) setlogLevelByte(LOG_LEVEL_WARNING);
		else if (property->equalsString(LOG_MODE_NOTICE)) setlogLevelByte(LOG_LEVEL_NOTICE);
		else if (property->equalsString(LOG_MODE_TRACE)) setlogLevelByte(LOG_LEVEL_TRACE);
		else if (property->equalsString(LOG_MODE_VERBOSE)) setlogLevelByte(LOG_LEVEL_VERBOSE);
		else setlogLevelByte(LOG_LEVEL_SILENT);
    }

	void onlogLevelWebChange(WProperty* property) {
		if (property->equalsString(LOG_MODE_SILENT)) setlogLevelWebByte(LOG_LEVEL_SILENT);
		else if (property->equalsString(LOG_MODE_FATAL)) setlogLevelWebByte(LOG_LEVEL_FATAL);
		else if (property->equalsString(LOG_MODE_ERROR)) setlogLevelWebByte(LOG_LEVEL_ERROR);
		else if (property->equalsString(LOG_MODE_WARNING)) setlogLevelWebByte(LOG_LEVEL_WARNING);
		else if (property->equalsString(LOG_MODE_NOTICE)) setlogLevelWebByte(LOG_LEVEL_NOTICE);
		else if (property->equalsString(LOG_MODE_TRACE)) setlogLevelWebByte(LOG_LEVEL_TRACE);
		else if (property->equalsString(LOG_MODE_VERBOSE)) setlogLevelWebByte(LOG_LEVEL_VERBOSE);
		else setlogLevelWebByte(LOG_LEVEL_SILENT);
	}

	void setlogLevelMqtt(byte lm) {
		const char * lms;
		lms=logLevelByteToString(lm);
		logLevel->setString(lms);
	}

	void setlogLevelWeb(byte lw) {
		const char * lws;
		lws=logLevelByteToString(lw);
		logLevelWeb->setString(lws);
	}

	const char * logLevelByteToString(byte lm) {
		if (lm == LOG_LEVEL_SILENT) return LOG_MODE_SILENT;
		else if (lm == LOG_LEVEL_FATAL) return LOG_MODE_FATAL;
		else if (lm == LOG_LEVEL_ERROR) return LOG_MODE_ERROR;
		else if (lm == LOG_LEVEL_WARNING) return LOG_MODE_WARNING;
		else if (lm == LOG_LEVEL_NOTICE) return LOG_MODE_NOTICE;
		else if (lm == LOG_LEVEL_TRACE) return LOG_MODE_TRACE;
		else if (lm == LOG_LEVEL_VERBOSE) return LOG_MODE_VERBOSE;
		else return LOG_MODE_SILENT;
    }

	void sendLog(int level, const char* message) {

		if (level<=getlogLevelWebByte()){
			wslog.printfAll(PSTR("%s: %s"), logLevelByteToString(level), message);
		}
		if (level<=getlogLevelByte()){
			String t = (String)network->getMqttTopic()+ "/" + MQTT_TELE +"/log" ;
			//network->log()->verbose(F("sendLog (%s)"), t.c_str());
			if (network->isMqttConnected()){
				network->publishMqtt(t.c_str(), ((String)logLevelByteToString(level)+": "+message).c_str());
			}
		}
	};

	

	void wsOnEvent(AsyncWebSocket * server, AsyncWebSocketClient * client, AwsEventType type, void * arg, uint8_t *data, size_t len){
		if (!client) return;
		if (client->status()!=WS_CONNECTED) return;
		if (!client->canSend()) return;
		if (type == WS_EVT_CONNECT){
			client->printf(PSTR("Hello WebSocket Client #%u\n"), client->id());
			client->printf(PSTR("Now you see all logs for logSeverity %s (or more important)\n"), logLevelWeb->c_str());
			client->ping();
		} else if (type == WS_EVT_DATA){
			AwsFrameInfo * info = (AwsFrameInfo*)arg;
			if (info->final && info->index == 0 && info->len == len){
				if (info->opcode == WS_TEXT){
					//client->text("I got your text message\n");
					data[len] = 0;
					//os_printf("%s\n", (char*)data);
				} else {
					//client->binary("I got your binary message\n");
				}
			}
			client->ping();
		} else if (type == WS_EVT_PONG){
		}
	}

void webserverInitHook(AsyncWebServer *webServer) {
	wslog.onEvent([this](AsyncWebSocket * server, AsyncWebSocketClient * client, AwsEventType type, void * arg, uint8_t *data, size_t len){
		this->wsOnEvent(server, client, type, arg, data, len);
	});
	webServer->addHandler(&wslog);
}
void webserverDeinitHook(AsyncWebServer *webServer) {
	wslog.closeAll();
}

private:
	WProperty* logLevel;
	WProperty* logLevelByte;
	WProperty* logLevelWeb;
	WProperty* logLevelWebByte;

};


#endif